### PR TITLE
Batched NNs for TorchANI

### PR DESCRIPTION
**Pull request by raimis**
A proof of concept to speed up the inference of ANI-like model by using batched matrix operations.

- [x] Implement `TorchANIBatchedNN`
- [x] Benchmarks
- [x] Tests
- [x] Documentation

See #11 for details

### CODE REVIEW

**Comment by raimis** [comment]
**Benchmarks**

Molecule: 46 atoms (`pytorch/molecules/2iuz_ligand.mol2`)
GPU: GTX 1080 Ti
Script: `nn/BenchmarkBatchedNN.py`

- ANI-2x (forward pass): 26 ms
- ANI-2x (forward & backward pass): 93 ms
- ANI-2x + BatchedNN (forward pass): 5.0 ms
- ANI-2x + BatchedNN (forward & backward pass): 15 ms

**Comment by raimis** [comment]
Additionally, the speed when using the faster symmetry functions (#5)

ANI-2x + #5 + BatchedNN (forward pass): 2.5 ms
ANI-2x + #5 + BatchedNN (forward & backward pass): 11 ms

**Comment by raimis** [comment]
Profiling of ANI-2x + #5 + BatchedNN (forward pass):
Screenshot from 2020-10-15 15-32-53

**Comment by raimis** [comment]
- In the profiler, it runs ~5.0 ms instead of 2.5 ms. So GPU are working more than it looks.
- Memory allocation for `ANISymmetryFunction` takes ~0.5 ms. We should get rid of this inefficiency.
- `ANISymmetryFunction` kernels take 0.2 ms.
- NN kernels runs 1.5 ms. For small molecules this will be a bottleneck eventually. It should be worth to look, if the auto-tuning TensorRT kernel can perform better. 
- At the `EnergyShifter` runs a lot of small kernel, which probably can be optimized too.

**Comment by peastman** [comment]
Very nice!  And it looks from the profile like there's still room for more improvement.

**Comment by raimis** [comment]
I have managed to speed up more the backward pass:
- ANI-2x + BatchedNN (forward pass): 5.2 ms
- ANI-2x + BatchedNN (forward & backward passes): 9.2 ms
- ANI-2x + #5 + BatchedNN (forward pass): 2.5 ms
- ANI-2x + #5 + BatchedNN (forward & backward passes): 5.1 ms

**Comment by raimis** [comment]
For some inexplicable reason, the backward pass performance drop, when converting to TorchScript:

**Comment by raimis** [comment]
With the tracing, it is slow too:
```python
nnp_tr = torch.jit.trace(nnp, ((species, positions),))
```

File: pytorch/TestBatchedNN.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+def test_model_serialization(deviceString, molFile):
+
+    device = torch.device(deviceString)
+
+    mol = mdtraj.load(f'molecules/{molFile}_ligand.mol2')
+    atomicNumbers = torch.tensor([[atom.element.atomic_number for atom in mol.top.atoms]], device=device)
+    atomicPositions = torch.tensor(mol.xyz, dtype=torch.float32, requires_grad=True, device=device)
+
+    nnp_ref = torchani.models.ANI2x(periodic_table_index=True).to(device)
+    nnp_ref.neural_networks = TorchANIBatchedNN(nnp_ref.species_converter, nnp_ref.neural_networks, atomicNumbers).to(device)
```
Comment (yueyericardo) [inline pytorch/TestBatchedNN.py:70]: Hi, thanks for the great work! 
For this line, should it be placed after `energy_ref` and `grad_ref` are calculated?

Comment (raimis) [reply pytorch/TestBatchedNN.py:70]: This is a test for a serialisation/deserialisation. The comparison with the original TorchANI is above.

Comment (yueyericardo) [reply pytorch/TestBatchedNN.py:70]: Ah, gotcha. Sorry for the misunderstanding!

**Comment by raimis** [comment]
I have solved the performance problem regarding TorchScript.

- ANI-2x + #5 + BatchedNN (forward pass): 2.6 ms
- ANI-2x + #5 + BatchedNN (forward & backward passes): 5.2 ms

**Comment by yueyericardo** [comment]
Hi Raimondas,

I found that GPU memory usage increase linearly as molecule size when initialize BatchedNN. 
Not sure whether it will be more expensive? Let me know if I did it wrong.

```
File: 1hz5.pdb, Molecule size: 973

species size: 100
  GPU Memory Used (nvidia-smi):  2242.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------

species size: 200
  GPU Memory Used (nvidia-smi):  3820.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------

species size: 300
  GPU Memory Used (nvidia-smi):  6032.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------

species size: 400
  GPU Memory Used (nvidia-smi):  7604.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------

species size: 500
  GPU Memory Used (nvidia-smi): 11542.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------

species size: 600
  GPU Memory Used (nvidia-smi): 13118.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------

species size: 700
  GPU Memory Used (nvidia-smi): 14694.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------

species size: 800
  GPU Memory Used (nvidia-smi):  9478.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------

species size: 900
  GPU Memory Used (nvidia-smi): 10530.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
----------------------------------------------------------------------
```

script to run it
```python
import os
import gc
import torch
import torchani
import pynvml
import numpy as np
from ase.io import read
from NNPOps.BatchedNN import TorchANIBatchedNN


def checkgpu(device=None):
    i = device if device else torch.cuda.current_device()
    real_i = int(os.environ['CUDA_VISIBLE_DEVICES'][0]) if 'CUDA_VISIBLE_DEVICES' in os.environ else i
    pynvml.nvmlInit()
    h = pynvml.nvmlDeviceGetHandleByIndex(real_i)
    info = pynvml.nvmlDeviceGetMemoryInfo(h)
    name = pynvml.nvmlDeviceGetName(h)
    print('  GPU Memory Used (nvidia-smi): {:7.1f}MB / {:.1f}MB ({})'.format(info.used / 1024 / 1024, info.total / 1024 / 1024, name.decode()))


file = '1hz5.pdb'
mol = read(file)
device = torch.device('cuda')
species_ = torch.tensor([mol.get_atomic_numbers()], device=device)
positions = torch.tensor([mol.get_positions()], dtype=torch.float32, requires_grad=False, device=device)
print(f'File: {file}, Molecule size: {species_.shape[-1]}\n')

for N in np.arange(100, 1000, 100):
    torch.cuda.empty_cache()
    gc.collect()
    species = species_[:, :N]
    print(f"species size: {species.shape[1]}")
    nnp = torchani.models.ANI2x(periodic_table_index=True, model_index=None).to(device)
    nnp.neural_networks = TorchANIBatchedNN(nnp.species_converter, nnp.neural_networks, species).to(device)
    checkgpu()
    print('-' * 70 + '\n')
```

pdb file at : https://raw.githubusercontent.com/yueyericardo/aev_benchmark/master/molecules/1hz5.pdb

**Comment by peastman** [comment]
That shows a large *decrease* in memory between 700 and 800?  This may not really be measuring what you think it is.

Try modifying the script so `N` is passed as a command line argument, then run the script repeatedly for different values.  That will remove uncertainty about when memory gets released.

**Comment by yueyericardo** [comment]
I know the memory might not be accurate.
But a directly run of species size of 900 is giving the similar result: 
```
File: 1hz5.pdb, Molecule size: 973

species size: 900
  GPU Memory Used (nvidia-smi): 10530.9MB / 16130.5MB (Tesla V100-PCIE-16GB)
```

**Comment by raimis** [comment]
@yueyericardo

The NN parameters are replicated for each atom to optimize memory layout for the batched multiplication. So, the memory increases linearly with a system size.

Effectively, this is trading memory for speed. For small molecules (~100 atoms), GPUs have more than enough of memory. For large molecules, we may need a different algorithm.

**Comment by peastman** [comment]
It seems like there ought to be a way to avoid having to duplicate them.  We just need to figure out how to get PyTorch to sum/broadcast correctly.  Only having one copy of the parameters should help caching efficiency, which would improve performance.

**Comment by raimis** [comment]
For small molecules, as far as I tested, it was the fastest algorithm, even though it wastes GPU memory and bandwidth, but it just one kernel launch.

For larger molecules, I guess if atoms are sorted by elements, the multiplication can be carried out for each element separately. This won't replicate the parameters, but would require more kernel launches.

**Comment by jchodera** [comment]
@raimis : Can this be updated and merged?

**Comment by raimis** [comment]
@peastman this is ready for a review!

**Review by peastman — APPROVED** [review]
Looks good!

**Pull request merged**